### PR TITLE
Add request rate limit to proxy

### LINKED ISSUE #168: Add request rate limit to proxy
**Blocked by #100**

To prevent abusing our public CORS proxy

- [ ] Allow only 500 request in a minute to the same domain from the same IP-address.
- [ ] Allow only 5000 request to any domain from the same IP-address.

On passing those limits:
1. Print to console (we will need it to debug)
2. Do not cancel response, just postpone (so client will not see an error, just slower connection).

These changes need to be done in `proxy/`. It will be cool to have tests, but if they are not stable or hard to make, we can avoid it.

### PR DESCRIPTION
**Pull request by janefawkes**
Fixes #168 

I sort of tested this by running the proxy and spamming it with batches of requests (like [this](https://gist.github.com/janefawkes/b3cf23c02d7295fc96cd31d9f214718c)). It works on my machine™, but `pnpm test` doesn't quite agree with it. 

Would love to hear any sort of feedback on this, because I'm not sure if I'm moving in the right direction or not.

Additional changes: proxy config and error handling have been divorced from the `createProxyServer` implementation.

### CODE REVIEW

File: proxy/proxy.ts [hunk 1]
```diff
@@ -13,17 +13,179 @@ class BadRequestError extends Error {
...
+    LIMIT: 500,
+    DURATION: 60 * 1000
+  },
+  GLOBAL: {
+    LIMIT: 5000,
+    DURATION: 60 * 1000
+  }
+}
+
+const delay = (ms: number): Promise<void> => {
```
Comment (ai) [inline proxy/proxy.ts:39]: Yoou can use:

```
import { setTimeout } from 'node:timers/promises`

await setTimeout(50)
```

File: proxy/proxy.ts [hunk 2]
```diff
@@ -13,17 +13,179 @@ class BadRequestError extends Error {
...
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+const rateLimitMap: Map<string, RateLimitInfo> = new Map()
+const requestQueue: Map<string, Promise<void>> = new Map()
+
+const isRateLimited = (ip: string, domain: string): boolean => {
```
Comment (ai) [inline proxy/proxy.ts:46]: We prefer `function isRateLimit() {}` instead of `const isRateLimit = () => {}`

File: proxy/proxy.ts [hunk 3]
```diff
@@ -13,17 +13,179 @@ class BadRequestError extends Error {
...
+    LIMIT: 5000,
+    DURATION: 60 * 1000
+  }
+}
+
+const delay = (ms: number): Promise<void> => {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+const rateLimitMap: Map<string, RateLimitInfo> = new Map()
```
Comment (ai) [inline proxy/proxy.ts:43]: All variables which is not `NAMED_LIKE_THIS` should be `let` according to our code style

File: proxy/proxy.ts [hunk 4]
```diff
@@ -13,17 +13,179 @@ class BadRequestError extends Error {
...
+
+  domainRateLimit.count += 1
+  globalRateLimit.count += 1
+  rateLimitMap.set(domainKey, domainRateLimit)
+  rateLimitMap.set(globalKey, globalRateLimit)
+
+  return false
+}
+
+const handleError = (e: any, res: ServerResponse): void => {
```
Comment (ai) [inline proxy/proxy.ts:87]:
```suggestion
const handleError = (e: unknown, res: ServerResponse): void => {
```

File: proxy/proxy.ts (hunk 3 above)
Comment (janefawkes) [reply proxy/proxy.ts:43]: > All variables which is not `NAMED_LIKE_THIS` should be `let` according to our code style

So I should not acknowledge constant references and use `const` only for real constants?

Comment (ai) [reply proxy/proxy.ts:43]: Yes, use `let` for variables (ESLint will tell you some cases when you run it)

File: proxy/proxy.ts (hunk 4 above)
Comment (janefawkes) [reply proxy/proxy.ts]: I guess [that](https://github.com/hplush/slowreader/pull/227/commits/729ee25fa33adede8dea6bcdeb110172d4bb6973) should do the job

File: proxy/proxy.ts [hunk 5]
```diff
@@ -13,16 +14,178 @@ class BadRequestError extends Error {
...
+    domainRateLimit.count = 0
+    domainRateLimit.timestamp = now
+  }
+
+  if (domainRateLimit.count >= RATE_LIMIT.PER_DOMAIN.LIMIT) {
+    return true
+  }
+
+  let globalKey = ip
+  let globalRateLimit = rateLimitMap.get(globalKey) || {
```
Comment (ai) [inline proxy/proxy.ts:62]: The logic is same between `globalKey` and `domainKey`.

Maybe we should have function like:

```
function isRateLimited(key: string, store: Map<string, RateLimitInfo>, limit: { LIMIT: number, DURATION: number })
```

**Comment by ai** [comment]
Now you need to fix tests (it is better to add some test for rate limit too) and we are ready to merge.

You can run tests by `cd proxy` and `pnpm test`.

You also need to fix ESLint warnings by calling `eslint --fix .`

**Comment by janefawkes** [comment]
ESLint warnings are fixed.

Not sure what to do with tests, though. I am a bit stuck with them and I don't really know how to ensure proper coverage
image

File: proxy/package.json [hunk 6]
```diff
@@ -8,7 +8,7 @@
   },
   "scripts": {
     "start": "tsx watch index.ts",
-    "test": "FORCE_COLOR=1 pnpm run /^test:/",
```
Comment (janefawkes) [inline proxy/package.json:11]: Changed it for the debug purposes, gonna add this line back in the next commit.

**Pull request closed without merging**